File: Others/100問/6_深さ優先探索/26_ABC138_D.cpp
#include<bits/stdc++.h>
#include <iostream>
#include <iomanip>
#include <cmath>

using namespace std;

#define ll long long
#define vec vector<int>
#define llvec vector<long long>

#define Graph vector<vector<int>>
#define rep(i,n) for (int i = 0; i < (n); ++i)
#define krep(i, k, n) for (int i = k; i < (int)(n); ++i)
#define rep_bit(bit,n) for (int bit = 0; bit < (1<<n); ++bit)
#define all(a) (a).begin(), (a).end()

// #include "/workspaces/AtCoder-cpp-env-main/cpp-dump/cpp-dump.hpp"
// namespace cp = cpp_dump;

vec seen;

void dfs(const Graph &G, vec &points,int node, int last_node){
  seen[node] = 1;
  points[node] += points[last_node];
  for (auto next_v: G[node]){
    if (seen[next_v]) continue;
    dfs(G, points, next_v, node);
  }
}


int main(){
  // input
  int N,Q; cin >> N >> Q;
  Graph graph(N+1);
  rep(i,N-1){
    int a,b;cin >> a >> b;
    graph[a].push_back(b);
    graph[b].push_back(a);
  }
  rep(i,N+1){
    sort(all(graph[i]));
  }
  // cpp_dump(graph);
  
  vec points(N+1,0);
  // cpp_dump(points);



  rep(i,Q){
    int p,x; cin >> p >> x;
    // cpp_dump(p,x);
    points[p] += x;
  }

  // cpp_dump(points);

  // init
  seen.assign(N+1,0);
  seen[0] = 1;

  // dfs
  int last_node = 0;
  int node = 1;
  dfs(graph, points, node, last_node);

  // cpp_dump(points);
  krep(i,1,N+1){
    cout << points[i] << " ";
  }
  cout << endl;

}

File: Others/100問/0_bk/26_ABC138_D.py
# 深さ優先探索（行きがけ）
import sys
input = sys.stdin.readline
sys.setrecursionlimit(10**7) #再帰関数の呼び出し制限

# グラフの作成(無向グラフでは#を消す)
N,Q = map(int, input().split())
graph = [[] for _ in range(N + 1)]
for _ in range(N-1):
    a,b = [int(x) for x in input().split()]
    graph[a].append(b)
    graph[b].append(a) # 無向グラフ
graph = [sorted(ls) for ls in graph]

points = [0] * (N+1)
for _ in range(Q):
    p,x = [int(x) for x in input().split()]
    points[p] += x


# 深さ優先探索
def dfs(node,seen,points):
    # print(node, seen)
    seen[node] = True
    for next_node in graph[node]:
        if seen[next_node]:
            continue
        else:
            points[next_node] = points[node] + points[next_node]
            dfs(next_node,seen,points)
    return points

seen = [False for _ in range(N+1)]
ans = dfs(1, seen, points)
# print(graph)
print(*ans[1:])

"""
4 3
1 2
1 3
2 4
"""